src/AbstractApiClient.php

```diff
@@ -0,0 +1,266 @@
+        $this->processQuery($query, $options);
+        $this->processHeaders($headers, $options);
+
+        try {
+            $response = $this->getClient()->send($request, $options);
+
+            return $this->processResponse($response, $request);
+        } catch (GuzzleException $e) {
+            $this->handleException($e, $request);
+        }
+    }
+
+    /**
+     * @param GuzzleException $e
+     * @param RequestInterface $request
+     * @throws BadResponseException
+     * @throws ErrorResponseException
+     * @throws TransportException
+     */
+    protected function handleException(GuzzleException $e, RequestInterface $request)
+    {
+        if ($e instanceof Exception\ClientException) { // 4xx
+            $this->processErrorResponse($e->getResponse(), $request);
+            $this->logError(
+                'Api unhandled [' . $e->getResponse()->getStatusCode() . '] Error Response from [' . $request->getUri() . ']',
+                $this->formatBadResponseException($e)
+            );
+
+            throw new BadResponseException($e->getMessage(), $e->getCode(), $e);
+        } elseif ($e instanceof Exception\BadResponseException) { // 4xx & 5xx
+            $this->logError(
+                'Api Bad Response from [' . $request->getUri() . '] Failed[' . $e->getResponse()->getStatusCode() . ']',
+                $this->formatBadResponseException($e)
+            );
+
+            throw new BadResponseException($e->getMessage(), $e->getCode(), $e);
+        } elseif ($e instanceof Exception\RequestException) {
+            $this->logError(
+                'Api problem with request to [' . $request->getUri() . ']',
+                $this->formatRequestException($e)
+            );
+        }
+
+        throw new TransportException($e->getMessage(), $e->getCode(), $e);
+    }
+
+    private function processQuery(array $query, array &$options): void
+    {
+        if (count($query) > 0) {
+            $options['query'] = $query;
+        }
+    }
+
+    private function processHeaders(array $headers, array &$options): void
+    {
+        $headers = array_merge($this->headers, $headers);
+
+        $headers['Content-Length'] = 0; // Issue https://github.com/guzzle/guzzle/issues/1645
+
+        if (count($headers) > 0) {
+            $options['headers'] = $headers;
+        }
+    }
+
+    private function formatBadResponseException(Exception\BadResponseException $e): array
+    {
+        return [
+            'message' => $e->getMessage(),
+            'request' => [
+                'headers'   => $e->getRequest()->getHeaders(),
+                'body'      => $e->getRequest()->getBody()->getContents(),
+                'method'    => $e->getRequest()->getMethod(),
+                'uri'       => $e->getRequest()->getUri(),
+            ],
+            'response' => [
+                'body'      => ($e->getResponse())?$e->getResponse()->getBody()->getContents():'[EMPTY]',
```

Comment:
Why `'[EMPTY]'`?

Reply:
If the response is `null` then it's just human-readable.